scripts/field_report.py

```diff
@@ -0,0 +1,262 @@
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import collections
+import re
+import sys
+from typing import Dict, List, Optional
+
+import clang.cindex #type: ignore
+from clang.cindex import Cursor, SourceLocation, TranslationUnit
+from clang.cindex import CursorKind, TypeKind
+
+from mbedtls_dev import typing_util
+
+
+class SanityCheck(Exception):
+    def __init__(self, msg: str) -> None:
+        super().__init__('Sanity check failed: ' + msg +
+                         '\nThis likely indicates a compilation error.' +
+                         '\nMaybe a missing include directory (-I)?')
+
+
+
+class FieldInfo:
+    """Information about a field of a structure."""
+
+    def __init__(self, node: Cursor) -> None:
+        self.node = node
+        self.lexical_uses = 0
+        self.size = node.type.get_size()
+        self.align = node.type.get_align()
+        # Empirically, offsetof is in bits, not bytes. To make the output
+        # easier to read, convert to bytes (the same unit as size and
+        # alignment), which means that bitfields will be located at their
+        # first byte.
+        self.offset = node.get_field_offsetof() // 8
+
+    def record_use(self, _node: Cursor) -> None:
+        self.lexical_uses += 1
+
+    def uses(self) -> int:
+        return self.lexical_uses
+
+    def is_indirect(self) -> bool:
+        """Whether access to this field is indirect on ARM Cortex-M0+."""
+        word_offset = self.offset // self.align
+        return word_offset >= 128
+
+    def score(self) -> int:
+        """A field's score is an estimate of its access cost.
+
+        The cost is calculated as the number of instructions needed to
+        access the field on ARM Cortex-M0+, multiplied by the number of uses
+        of the field inside the library.
+        """
+        cost_per_use = 1 + 2 * self.is_indirect()
+        return self.uses() * cost_per_use
+
+
+class Ast:
+    """Abstract representation of the source code."""
+
+    def __init__(self, options) -> None:
+        """Prepare for analysis of C source files."""
+        if options.clang_library_file:
+            clang.cindex.Config.set_library_file(options.clang_library_file)
+        self.parse_options = []
+        if options.target:
+            self.parse_options += ['-target', options.target]
+        for d in options.include:
+            self.parse_options.append('-I' + d)
+        for d in options.define:
+            self.parse_options.append('-D' + d)
+        self.index = clang.cindex.Index.create()
+        self.files = {} #type: Dict[str, TranslationUnit]
+        # fields[TYPE_OR_STRUCT_NAME][FIELD_NAME]
+        self.fields = {} #type: Dict[str, Dict[str, FieldInfo]]
```

Comment:
The way this data is accessed makes it feel like it should be a `DataClass()` [ref](https://docs.python.org/3/library/dataclasses.html)

We have the data stored, and could contain methods to perfom the following function

- Getter method to return dictionary sorted by field_name or by entry order.
- Boolean method to check for SanityCheck 
- An iterative function to printout the data like report_field, pprint or __str__ , __repr__ overrides.

Reply:
`dataclasses` was added in Python 3.7. We still officially support 3.6 and run CI on 3.5.